Improved shapes

ghislaineguerin opened the pull request.

ghislaineguerin commented:
Update:

A white frame and border are now used to enhance mobile view. This will simplify the responsiveness of most shapes.

![localhost_4200_ (1)](https://user-images.githubusercontent.com/845767/59615121-4f7cb800-9122-11e9-8322-f9554aa2c1f3.png)
![localhost_4200_](https://user-images.githubusercontent.com/845767/59615123-4f7cb800-9122-11e9-8f0a-0bc6bdc7efe1.png)

marcoow commented:
> A white frame and border are now used to enhance mobile view. This will simplify the responsiveness of most shapes.

Hm, is this what you agreed on? I'm not a huge fan tbh - I think it should be possible to implement the design we have on abstract.com?

ghislaineguerin commented:
@marcoow I've added the white frame to simplify the implementation. Florian was away. I have now fixed most issues so I think I can focus on mobile. I will fix it to look like the abstract version.

marcoow commented:
Hm, I'd say let's try and simplify the implementation without changing the design (unless that turns out to be impossible for some reason of course). Do you think that makes sense?

ghislaineguerin commented:
@marcoow Sure, I can do that.

ghislaineguerin commented:
@pichfl looks great, mobile it's also working fine for me.

marcoow commented on src/ui/styles/blocks/offset.block.css:
```diff
@@ -55,3 +55,5 @@
   margin-bottom: 21rem;
 }
+
+
```
👀

pichfl commented on src/ui/components/PageHomepage/stylesheet.css:
```diff
@@ -4,6 +4,7 @@
 @block offset from "../../styles/blocks/offset.block.css";
 @block typography from "../../styles/blocks/typography.block.css";
 
+
```
Took me until now to realize that prettier isn't run for CSS files, added PR #606 for that. Will rebase once merged.

pichfl commented on src/ui/components/ShapeAcross/component.ts:
```diff
@@ -1,7 +1,23 @@
 import ShapeBase from '../ShapeBase/component';
 
 export default class ShapeAcross extends ShapeBase {
+  public maskId = null;
```
No reason not to do it.

marcoow replied:
not sure what you mean

pichfl replied:
Sorry. Meant to say: Will change to private and update similar props in other components as well.

pichfl replied:
Actually, if I make these private, I get warnings in TSLint about props that are never read, as the linter doesn't know about the templates. Should I still do that and add a comment to disable that warning?

marcoow replied:
I'd say let's just not specify `private`/`public` at all.

pichfl replied:
It gets inserted by the formatter task automatically

pichfl commented on src/ui/components/ShapeAcross/component.ts:
```diff
@@ -1,7 +1,23 @@
+
+    this.maskId = Math.random()
+      .toString(36)
+      .substr(2);
```
This is only the latest implement of that code, it's used all across the components. 
Sadly I can not use a static string for those ids as they would clash if a component is reused twice on a page, which can and does happen.

It's used in `ShapeBase` and any Shape* class that has inline SVG references. 

I was hoping that a changing id wouldn't actually have any impact on the pre-render as they are used only locally in that component, which in turn makes a strong argument for changing those fields to private.

marcoow replied:
Yeah, only recognized that now - we should have a separate issue to handle this. Can you create one?

pichfl replied:
Done, see #608

pichfl commented on src/ui/components/ShapeBold/component.ts:
```diff
@@ -0,0 +1,21 @@
+      style.set('--shape-color-b', this.args.backgroundB);
+    }
+
+    return [...style.entries()].map(([key, value]) => `${key}:${value}`).join(';');
```
Overengineering at it's finest. Will replace with a less computationally offensive solution. There is absolutely no need for the `Map` in the first place.

Sidennote: The `[...style.entries()].map` call is not used for uniqueness, but as the only way to map the entries inside a Map, since the Map object does not support `.map()` itself.

pichfl commented on src/ui/components/ShapeFeature/component.ts:
```diff
@@ -18,4 +18,14 @@ export default class ShapeFeature extends ShapeBase {
+    }
+
+    return style;
+  }
```
Tried to keep the method signature similar to the code above, I'm not attached to that bit, will refactor.

marcoow commented on src/ui/components/ShapeAcross/component.ts:
```diff
@@ -1,7 +1,23 @@
+      style += `--shape-color: ${this.args.background}`;
+    }
+
+    return style;
```
I'd simplify the method to

```es6
get backgroundStyle() {
  if (this.args.background) {
    return `--shape-color: ${this.args.background}`;
  }
}
```

pichfl commented:
Everything is compressed with ImageOption and where possible with ImageAlpha as well. For the overlapping images we have to rely on pngs (at least for stable solution) so while unfortunate, there is no other option here 😞

marcoow commented:
@pichfl can we check we can shrink the image dimensions a bit maybe? 350KB is pretty huge…
